gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/GlutenParquetRowIndexSuite.scala

```diff
@@ -16,6 +16,337 @@
+
 import org.apache.spark.sql.GlutenSQLTestsBaseTrait
+import org.apache.spark.sql.execution.FileSourceScanExec
+import org.apache.spark.sql.execution.datasources.FileFormat
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.execution.datasources.v2.parquet.ParquetDataSourceV2
+import org.apache.spark.sql.functions.{col, max, min}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{LongType, StringType}
+
+import org.apache.hadoop.fs.Path
+import org.apache.parquet.column.ParquetProperties._
+import org.apache.parquet.format.converter.ParquetMetadataConverter
+import org.apache.parquet.hadoop.ParquetOutputFormat
+import org.apache.parquet.hadoop.ParquetWriter.DEFAULT_BLOCK_SIZE
+
+import java.io.File
+
+import scala.collection.JavaConverters._
+
+class GlutenParquetRowIndexSuite extends ParquetRowIndexSuite with GlutenSQLTestsBaseTrait {
+  import testImplicits._
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    sparkContext.setLogLevel("info")
+  }
+
+  private def readRowGroupRowCounts(path: String): Seq[Long] = {
+    ParquetFooterReader
+      .readFooter(
+        spark.sessionState.newHadoopConf(),
+        new Path(path),
+        ParquetMetadataConverter.NO_FILTER)
+      .getBlocks
+      .asScala
+      .map(_.getRowCount)
+  }
+
+  private def readRowGroupRowCounts(dir: File): Seq[Seq[Long]] = {
+    assert(dir.isDirectory)
+    dir
+      .listFiles()
+      .filter(f => f.isFile && f.getName.endsWith("parquet"))
+      .map(f => readRowGroupRowCounts(f.getAbsolutePath))
+  }
+
+  /** Do the files contain exactly one row group? */
+  private def assertOneRowGroup(dir: File): Unit = {
+    readRowGroupRowCounts(dir).foreach {
+      rcs => assert(rcs.length == 1, "expected one row group per file")
+    }
+  }
+
+  /**
+   * Do the files have a good layout to test row group skipping (both range metadata filter, and by
+   * using min/max).
+   */
+  private def assertTinyRowGroups(dir: File): Unit = {
+    readRowGroupRowCounts(dir).foreach {
+      rcs =>
+        assert(rcs.length > 1, "expected multiple row groups per file")
+        assert(rcs.last <= DEFAULT_MINIMUM_RECORD_COUNT_FOR_CHECK)
+        assert(
+          rcs.reverse.tail.distinct == Seq(DEFAULT_MINIMUM_RECORD_COUNT_FOR_CHECK),
+          "expected row groups with minimal row count")
+    }
+  }
+
+  /**
+   * Do the files have a good layout to test a combination of page skipping and row group skipping?
+   */
+  private def assertIntermediateRowGroups(dir: File): Unit = {
+    readRowGroupRowCounts(dir).foreach {
+      rcs =>
+        assert(rcs.length >= 3, "expected at least 3 row groups per file")
+        rcs.reverse.tail.foreach {
+          rc =>
+            assert(
+              rc > DEFAULT_MINIMUM_RECORD_COUNT_FOR_CHECK,
+              "expected row groups larger than minimal row count")
+        }
+    }
+  }
+
+  case class GlutenRowIndexTestConf(
```

Comment:
@gaoyangxiaozhu Can we leave a TODO to remove this change after https://github.com/facebookincubator/velox/pull/8864 merged?

Reply:
@gaoyangxiaozhu  please open an issue for above. Thanks.

Reply:
@JkSelf even 8864 merged, for 3.4/3.5 we still can't remove this chagne right ? Since I only see it been set here https://github.com/apache/incubator-gluten/blob/37be4aa026c0aa9265a89fb7ae6b4210a1eaafed/cpp/velox/operators/writer/VeloxParquetDatasource.cc#L101

but looks `veloxParquetDatasource.cc` is not used for 3.4/3.5 parquet write ?

Reply:
@zhli1142015 and @JkSelf creating a issue to track https://github.com/apache/incubator-gluten/issues/5751

Reply:
@gaoyangxiaozhu Yes, 3.4 will not reach the code of V[VeloxParquetDatasource.cc](https://github.com/apache/incubator-gluten/blob/37be4aa026c0aa9265a89fb7ae6b4210a1eaafed/cpp/velox/operators/writer/VeloxParquetDatasource.cc#L101). However, with the https://github.com/facebookincubator/velox/pull/8864, we can set the flush policy where the TableWriteNode is created https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlan.cc#L583. Currently, this does not support pageSize, but we can add it in Velox in a similar manner."

Reply:
@gaoyangxiaozhu It appears that [VeloxParquetDatasource.cc](https://github.com/apache/incubator-gluten/blob/37be4aa026c0aa9265a89fb7ae6b4210a1eaafed/cpp/velox/operators/writer/VeloxParquetDatasource.cc#L101) does not set the pageSize either. How is it that this unit test can pass in both Spark 3.2 and Spark 3.3? Perhaps I miss something. Thanks.